Partial application of non-curried functions

pigumar1 opened the pull request:
Continuation of #957.
This feature introduces partial application of non-curried functions. As #514 mentioned:

> During elaboration, the partial application is replaced with a new function literal that takes in the unapplied arguments, then inserts them where they belong in the application of the original function.

The partial application of non-curried functions has been mostly implemented in #957, but it violates the principle that "deferral expressions should not synthesize to or analyze against any type." Additionally, there are some design issues to address. The purpose of this pull request is to resolve these issues.

### Examples

> ```
> val f: ('a, 'b, 'c) → 'd
> val x: 'a
> val y: 'b
> val z: 'c
> ```
> 
> |External Expression|Internal Expression|Synthesized Type|
> |:--|:--|:--|
> |`f (x, y, z)`|`f (x, y, z)`|`'d`|
> |`f (x, _, z)`|`λb.{f (x, b, z)}`|`'b → 'd`|
> |`f (_, y, _)`|`λ(a, c).{f (a, y, c)}`|`('a, 'c) → 'd`|
> |`f (_, _, _)`|`λ(a, b, c).{f (a, b, c)}`|`('a, 'b, 'c) → 'd`|

A function that has an input type inconsistent with a non-trivial product type cannot be partially applied. `(fun x : Int -> x)(_)` or `(fun x : () -> x)(_)` should result in some kind of error that prompts the user to work on the function rather than meaningless partial application.

If the input type of the function is a type hole, we cannot statically analyze whether the function can be partially applied. This means that the expression `(fun x -> x + 1)(_, 1)` is statically valid, and would be elaborated as `λy.{(λx.{x + 1}) (y, 1)}`. 

Providing a deferral singleton as an argument to such functions, like `(fun x -> x + 1)(_)`, is something programmers can do. There are different ways to handle such expressions statically:
* We can enforce that the argument for partial application must be a non-empty tuple, and then have this kind of expression result in an error to alert programmers about "meaningless partial application."
* Alternatively, we can refrain from imposing restrictions on the argument for partial application, allowing such expressions to be valid from a static perspective and be elaborated accordingly.
* ~~A more extreme approach would be to further restrict the input type of the function, disallowing type holes as valid input types.~~

### Implementation
* Add two new `UExp.term`s: `Deferral` and `DeferredAp(UExp.t, UExp.t)`.
* `_` introduces a `Deferral` expression. When a `Deferral` is either **a projection of the argument of a partial applicaton**, or **the argument of a partial application itself**, it's interpreted as bounded. Any other case would result in a `FreeDeferral` error. Adding extra parentheses around `_` would also result in a `FreeDeferral` error.
* Add a new `Info.error_exp` variant called `InconsistentWithDeferrableArrow` for functions that cannot be partially applied. ~~Assigning this error as Info to a function expression is challenging. Similar to the dilemma encountered in #957, although the current implementation of `Info.exp` includes `status_exp` to allow bypassing the function `status_exp` and directly assigning `status_exp` to an expression in `Statics.uexp_to_info_map`, the function `status_exp` is also used in other places like `elaborator.re`, resulting in inconsistent `status_exp` for the same expression in different locations. It would be best to restrict the use of the function `status_exp` to `Statics.re` only, but if that is not possible, to maintain data consistency, it would still require implementing the assignment of the new error in `status_exp`. Then,~~ a new variant of `Mode.t` called `SynDeferrableFun` would be needed, which, in addition to requiring the expression type to be consistent with the arrow type in `SynFun` mode, would also require its input type to be consistent with a non-trivial product type.
* If a function cannot be partially applied, it will match the arrow type `? -> ?`. If a function can be partially applied, the argument expression (arguments applied and deferrals) may not match the input type of the function. The verification process involves analyzing each element of the argument expression against the corresponding function input type, in which `Deferral`s are skipped.
* If there is a type inconsistency between the input and the argument types, a new variant of `Info.error_exp` called `InconsistentPartialApArg` will be assigned to the argument. In this error, `Deferral`s do not synthesize to or analyze against any type and appear only as placeholders (`_`) in the error message. ~~Similar to the challenges encountered with assigning `InconsistentWithDeferrableArrow`, if we want to implement error assignment within the function `status_exp`, the information about the *term* must be part of the input to `status_exp`. A better solution is to implement this within `Statics.uexp_to_info_map`, which currently leads to data inconsistency.~~
* The entire partial application is synthesized as an arrow type. If there is a type inconsistency between the input and the argument types, the input type of the arrow type is `Unknown`. In any case, the output type of the arrow type remains the same as the output type of the original function expression, providing as much as possible information to the programmer.
* In `elaborator.re`, when a partial application can be performed normally, the deferral expressions act as placeholders and are not elaborated. When there is a static error, partial application cannot be performed. To provide more reference information to programmers, a **full application** will be performed instead. This means that deferral expressions will be further elaborated into `EmptyHole`.
```
let plus = (fun (x, y) -> x + y) in
```
|External Expression|Internal Expression|Synthesized Type|
|:--|:--|:--|
|`_`|`InvalidText("_")`|`?`|
|`1 (_)`|`1 (?)`|`?`|
|`plus (_)`|`plus (?)`|`?`|
|`plus (_, 1)`|`(fun ~deferred0 -> plus(~deferred0, 1))`|`? -> Int`|
|`plus ((_), 1)`|`plus ((?), 1)`|`?`|
|`plus (_, _)`|`(fun (~deferred0, ~deferred1) -> plus(~deferred0, ~deferred1))`|`(?, ?) -> Int`|
|`plus (_, _, 1)`|`plus (?, ?, 1)`|`?`|

pigumar1 committed:
added new terms

pigumar1 committed:
fixed new terms

pigumar1 committed:
Update MakeTerm.re

pigumar1 committed:
Merge branch 'adt-defs-after' into haz3l-partial-application-no-self-type

pigumar1 committed:
Added status_deferral

pigumar1 committed:
Added new error_exp variant

TODO: assigning this error as Info to a function expression

pigumar1 committed:
Added FreeDeferral + other

1. Added FreeDeferral
2. Changed the implementation of Deferral

pigumar1 committed:
MakeTerm

pigumar1 committed:
formatted

pigumar1 committed:
Runnable

pigumar1 commented:
Based on yesterday's discussion, I have made some revisions to the parsing of `DeferredAp` so that parentheses, commas, and deferral expressions are all part of the `DeferredAp` syntax:
![image](https://github.com/hazelgrove/hazel/assets/108375845/894ac749-0269-46f0-a798-65854aead0a4)
![image](https://github.com/hazelgrove/hazel/assets/108375845/e80c5edd-9014-46cd-9659-9c74030a5739)
![image](https://github.com/hazelgrove/hazel/assets/108375845/cefd1011-64f8-40aa-beaa-fd921e7afbbf)
![image](https://github.com/hazelgrove/hazel/assets/108375845/8a718715-2f86-4b09-822e-bf0eb604d451)

The `InconsistentPartialApArg` will result in the following behavior. In `Statics.uexp_to_info_map`, an inconsistent partial application is assigned the self type `IsInconsistentPartialApArg`, which is interpreted as `InconsistentPartialApArg` in any mode to avoid deferrals' type checking. 
The current implementation of partial application still retains its tuple argument. Because this tuple argument shares the same syntax as the partial application itself, and the partial application takes precedence in parsing, it is not possible to assign this error to the argument separately.
![image](https://github.com/hazelgrove/hazel/assets/108375845/500234a3-24b7-43fd-865f-941894b47dd5)

The issue I'm currently facing is that all syntax elements of the partial application point to the same `Info`. Consequently, if there is a static error, the error message and type information cannot be displayed simultaneously.

* Would it be beneficial to split the partial application into two parts when encountering `InconsistentPartialApArg`? The first part would include **deferrals, comma tokens, and tuple elements**, while the second part would consist of **the parentheses token and the function expression**. This approach would resemble the handling of full application, where selecting the **parentheses** provides the **type** of the application expression, and selecting the **commas** provides **error information**. This splitting would occur only when `InconsistentPartialApArg` occurs, so when there are no static errors, the "type" of the argument remains hidden, ensuring that deferrals do not need to synthesize or analyze against specific types.
* Another approach is to fully **destructure the tuple argument during parsing**. This approach may be easier to implement and syntactically more intuitive. However, it would result in fewer and less accurate information being provided to programmers, and it would not allow the error message and type information to be displayed simultaneously.

pigumar1 commented:
Also, this design issue needs to be resolved:
> Providing a deferral singleton as an argument to such functions, like (fun x -> x + 1)(_), is something programmers can do. 
>
> There are different ways to handle such expressions statically:
> 
> * We can enforce that the argument for partial application must be a non-empty tuple, and then have this kind of expression result in an error to alert programmers about "meaningless partial application."
> * Alternatively, we can refrain from imposing restrictions on the argument for partial application, allowing such expressions to be valid from a static perspective and be elaborated accordingly.

pigumar1 commented:
~~Furthermore, now that deferrals have become part of the partial application syntax, selecting a deferral will provide information about the partial application rather than the deferral itself. In that case, would it be better to treat unbounded deferrals as `BadToken`s as well? This way, there would be no independent deferral syntax form anymore.~~

Another question is whether we should enforce the presence of at least one non-deferral argument in a partial application, otherwise generating an error. This may facilitate handling certain edge cases.

pigumar1 committed:
InconsistentPartialApArg error_exp

1. InconsistentPartialApArg error_exp
2. IsInconsistentPartialApArg Self.exp
3. Argument statics

cyrus- commented:
I think it's fine if the error message is all you can see when there is a static error and your cursor is on the parens or commas of a deferral.

I think making it an error to have only one deferral with no other arguments makes sense, that is never what you want.

(If you want to chat a bit more after lab meeting today let me know.)

pigumar1 committed:
New implementation

pigumar1 committed:
Update documets

pigumar1 committed:
Fixed small bugs

For input-type-is-non-prod functions (e.g.: Int -> Int and () -> Int), deferral singleton argument satisfies expected type

pigumar1 commented:
When there is no static error, selecting `_` in a partial application displays the type of the entire partial application instead of the type matched by the deferral, which is misleading. As a result, `_` is no longer part of the partial application syntax but exists as an independent deferral expression.
![image](https://github.com/hazelgrove/hazel/assets/108375845/367a74a9-1769-457a-9484-092a652bd687)

Furthermore, we have added a new restriction to partial application: the argument must include at least one non-deferral actual value; otherwise, the error `MeaninglessPartialAp` will be assigned to the entire partial application.

Finally, if possible, it would be preferable to assign a separate `ok_exp` status to ok deferral expressions. The current two `ok_exp` statuses - `SynConsistent` and `AnaConsistent` - both involve inappropriate type checking of deferral expressions, which are sometimes displayed in the cursor inspector. The **bold information** below shows examples of these inappropriate type checkings. A separate `ok_exp` status would allow us to bypass these issues in `status_exp`.

##
![image](https://github.com/hazelgrove/hazel/assets/108375845/fadd6504-8e5c-473d-975a-e5e5f48e7c46)

#### Case 1
![image](https://github.com/hazelgrove/hazel/assets/108375845/a8c12d5e-992d-4aba-b49a-c8bd07768395)
* Error message: "~~Deferral does not appear in a partial application as a placeholder~~Deferral must appear as a function argument"

#### Case 2 (Meaningless partial application evaluates to its function expression)
![image](https://github.com/hazelgrove/hazel/assets/108375845/d96b677b-af47-45f8-b792-c49d65d14e42)
* Select the parentheses: "~~Meaningless partial application: expected at least one non-deferral expression~~Expected at least one non-deferred argument"
* Select the deferral: "~~**[EXP] has expected & actual type ?**~~[EXP] satisfies expected type ?"

#### Case 3
![image](https://github.com/hazelgrove/hazel/assets/108375845/30c075b0-ab53-4fa4-b987-261394d902d0)
![image](https://github.com/hazelgrove/hazel/assets/108375845/0e8becec-dbdf-468c-b7bc-aa5254e1e2df)
* Select the parentheses or the comma: "[EXP] has type ? -> ?"
* Select the deferral: "~~**[EXP] has expected & actual type ?**~~[EXP] satisfies expected type ?"

#### Case 4 | 5 | 6
![image](https://github.com/hazelgrove/hazel/assets/108375845/723d188f-83b9-40e1-b746-ae88dc1274ee)
![image](https://github.com/hazelgrove/hazel/assets/108375845/daabdb5c-c40b-4dcc-999d-686d2187b9aa)
* Select the parentheses of the partial application: "~~Meaningless partial application: expected at least one non-deferral expression~~Expected at least one non-deferred argument"
* Select the deferral: "[EXP] satisfies expected type Int"

![image](https://github.com/hazelgrove/hazel/assets/108375845/d8ac4533-a39e-484d-8a53-61a93204c37a)
* Select the deferral: "[EXP] satisfies expected type ()"

![image](https://github.com/hazelgrove/hazel/assets/108375845/af9ef514-29a8-419a-acde-5dad8b1bbc87)
* Select the deferral: "~~**[EXP] has expected & actual type ?**~~[EXP] satisfies expected type ?"

#### Case 7
![image](https://github.com/hazelgrove/hazel/assets/108375845/3545dbe3-6bc6-4108-aa53-5b953df47741)
![image](https://github.com/hazelgrove/hazel/assets/108375845/1ef619a8-259c-4ba2-bc2a-1642ab04485f)
* Select the parentheses (or the comma) of the partial application: "Arity mismatched partial application: expected 1 ~~expression~~argument, ~~found~~got 2"
* Select the deferral: "~~**[EXP] has expected & actual type ?**~~[EXP] satisfies expected type ?" (consistent with the handling of a full application when there is an "arity error")

#### Case 8
![image](https://github.com/hazelgrove/hazel/assets/108375845/d8bbccb6-49a5-46fb-b833-a705ee688918)
![image](https://github.com/hazelgrove/hazel/assets/108375845/f1acf7c9-d081-4879-9f59-7c542a31d35f)
* Select the parentheses (or the comma) of the partial application: "[EXP] has type ? -> ?"
* Select the deferral: "~~**[EXP] has expected & actual type ?**~~[EXP] satisfies expected type ?"

#### Case 9
![image](https://github.com/hazelgrove/hazel/assets/108375845/608bf8ba-0590-4476-9747-c04fb8168348)
* Select the parentheses (`InconsistentPartialAp` has higher priority): "Arity mismatched partial application: expected 2 ~~expressions~~arguments, ~~found~~got 1"
* Select the deferral: "~~**[EXP] has expected & actual type ?**~~[EXP] satisfies expected type ?" ~~or "**[EXP] satisfies expected type (?, ?)**"~~ ~~(this piece of information might be misleading. I'm considering assigning an error to the deferral as well. To do so, we need to have a self type for the deferral, which will be specially treated in `status_exp`; if not, either the deferral would have the self type `Common(Just(Unknown(Internal)))`, which is consistent with the function's input type, or we need to bypass `status_exp` to assign the error.)~~

#### Case 10 | 11 | 12
![image](https://github.com/hazelgrove/hazel/assets/108375845/16f8d7d0-ae02-4a81-bace-e1d29c9167d6)
![image](https://github.com/hazelgrove/hazel/assets/108375845/b8254853-4e4f-402c-95e3-72d988291099)
* Select the parentheses or the comma: "[EXP] has type Int -> Int"
* Select the deferral: "[EXP] satisfies expected type Int"

![image](https://github.com/hazelgrove/hazel/assets/108375845/e1299647-6146-499d-87bb-07757de25f61)
* Select the deferral: "~~**[EXP] has expected & actual type ?**~~[EXP] satisfies expected type ?"

![image](https://github.com/hazelgrove/hazel/assets/108375845/be0f165f-f829-42c8-ac28-825b7e392ec0)
![image](https://github.com/hazelgrove/hazel/assets/108375845/2ae528a8-9fd8-4d2d-bf02-761ece2c91b0)

#### Case 13 (This is not a partial application)
![image](https://github.com/hazelgrove/hazel/assets/108375845/bbfc2fe7-48df-4721-a915-6471723cd88e)
* Select the deferral: "~~Deferral does not appear in a partial application as a placeholder~~Deferral must appear as a function argument"

#### Case 14
![image](https://github.com/hazelgrove/hazel/assets/108375845/6fd6378f-a337-46d6-9a82-212cdd566788)
* Select the parentheses or the comma: "Arity mismatched partial application: expected 2 ~~expressions~~arguments, ~~found~~got 3"
* Select the deferrals: "~~**[EXP] has expected & actual type ?**~~[EXP] satisfies expected type ?" (consistent with the handling of a full application when there is an "arity error")

#### Case 15
![image](https://github.com/hazelgrove/hazel/assets/108375845/236d8eec-dd58-4c5a-aa48-9353ca8dbf69)
![image](https://github.com/hazelgrove/hazel/assets/108375845/691d7f29-ac02-420f-89ea-15b75dcf85af)

pigumar1 committed:
Arity mismatched partial application

pigumar1 committed:
Reformat + ErroneousPartialAp

pigumar1 committed:
Elaborator + small fixes

cyrus- commented on src/haz3lweb/LangDocMessages.re:
```diff
@@ -164,6 +165,27 @@ let triv_exp: form = {
+let deferral_exp_group = "deferral_exp_group";
+let deferral_exp: form = {
+  let explanation = {
+    message: "Deferral expression. This marks a placeholder for some part of the argument expression in a partial application whose value supply is deferred to future applications.",
```
"Deferral expression. This marks an argument that has not yet been applied in a partial application."

cyrus- commented on src/haz3lweb/LangDocMessages.re:
```diff
@@ -164,6 +165,27 @@ let triv_exp: form = {
+  };
+  let deferral_exp_ex = {
+    sub_id: "deferral_exp_ex",
+    term: mk_example("let plus1 =\n(fun (x, y) -> x + y)(_, 1)\nin plus1(2)"),
```
this example goes off-screen and is too convoluted. try 

```
let plus = fun (x, y) -> x + y in 
let incr = plus(_, 1) in 
incr(5)
```

cyrus- commented on src/haz3lweb/LangDocMessages.re:
```diff
@@ -1541,6 +1564,12 @@ let conapp_exp_ex = {
+let deferred_funapp_exp_ex = {
+  sub_id: "deferred_funapp_exp_ex",
+  term: mk_example("(fun (x, y) -> x + y)(_, 1)"),
+  message: "The plus function is partially applied to (_, 1). Part of the tuple pattern y is bound to 1 in the function body and the whole partial application evaluates to the new function plus1.",
```
this message doesn't seem to be showing up correctly

cyrus- commented on src/haz3lweb/LangDocMessages.re:
```diff
@@ -1581,6 +1610,44 @@ let conapp_exp: form = {
+];
+let deferred_funapp_exp: form = {
+  let explanation = {
+    message: "Partial function application. Bound part(s) of the [*function*](%i) input pattern to supplied values in the [*argument*](%i).",
```
this message doesn't seem to be showing up correctly. just get the normal function application docs.

pigumar1 replied:
~~I managed to let the message show up in my local build, but it doesn't work correctly on the deployed build for unknown reasons... Here comes a screenshot of the message:~~
![image](https://github.com/hazelgrove/hazel/assets/108375845/f348ef06-3c16-4aec-bff0-a86bcd8d23a1)

cyrus- commented on src/haz3lweb/view/CursorInspector.re:
```diff
@@ -154,6 +154,29 @@ let typ_err_view = (ok: Info.error_typ) =>
+  | InHole(FreeDeferral) =>
+    div_err([
+      text(
+        "Deferral does not appear in a partial application as a placeholder",
```
"Deferral must appear as a function argument."

cyrus- commented on src/haz3lweb/view/CursorInspector.re:
```diff
@@ -154,6 +154,29 @@ let typ_err_view = (ok: Info.error_typ) =>
+  | InHole(ErroneousPartialAp(Meaningless)) =>
+    div_err([
+      text(
+        "Meaningless partial application: expected at least one non-deferral expression",
```
"Expected at least one non-deferred argument."

cyrus- commented on src/haz3lweb/view/CursorInspector.re:
```diff
@@ -154,6 +154,29 @@ let typ_err_view = (ok: Info.error_typ) =>
+      text(
+        "Arity mismatched partial application: expected "
+        ++ string_of_int(expected)
+        ++ " expression"
```
"argument(s)"

cyrus- commented on src/haz3lweb/view/CursorInspector.re:
```diff
@@ -154,6 +154,29 @@ let typ_err_view = (ok: Info.error_typ) =>
+        ++ string_of_int(expected)
+        ++ " expression"
+        ++ (expected == 1 ? "" : "s")
+        ++ ", found "
```
got

cyrus- reviewed (changes requested):
- there is a merge conflict with `dev`
- the decoration when the cursor is on a comma in a partial application are missing the `f` (it works correctly when on the parentheses):
![image](https://github.com/hazelgrove/hazel/assets/280638/a0f6f5f2-bc01-48b5-ad72-9453a8217bf4)

pigumar1 committed:
Small fixes + new ok_exp + new deferral self & term base

pigumar1 committed:
Document updates

pigumar1 committed:
Merge branch 'adt-defs-after' into haz3l-partial-application-no-self-type

pigumar1 committed:
Update TypBase.re

pigumar1 committed:
Document for DeferredAp

pigumar1 committed:
Code optimization

pigumar1 committed:
Code optimization 2

pigumar1 commented on src/haz3lcore/dynamics/Elaborator.re:
```diff
@@ -261,6 +262,48 @@ let rec dhexp_of_uexp =
+          let* ty_fn = fixed_exp_typ(m, fn);
+          let (ty_arg, ty_ret) = Typ.matched_arrow(ty_fn);
+          let ty_ins = Typ.matched_args(List.length(args), ty_arg);
+          /* Substitute all deferrals for new variables */
```
> there's no actual data dependency between elements here (i think) so instead of folding, just map, and then use OptUtil.sequence (see other examples in this file)

I gave it a try. The issue I encountered is that lines 206-207 require both pats and d_args to share the same name, and this name is determined by the length of the growing pats, so perhaps using a fold would be more suitable.

disconcision replied:
Makes sense

pigumar1 committed:
Fixed decoration bug

Co-Authored-By: Haoxiang Fei <tonyfettes@tonyfettes.com>

pigumar1 commented on src/haz3lweb/view/Deco.re:
```diff
@@ -115,7 +115,8 @@ module Deco =
       let range: option((Measured.Point.t, Measured.Point.t)) = {
         // if (Piece.has_ends(p)) {
-        switch (TermRanges.find_opt(Piece.id(p), M.term_ranges)) {
+        let id = Id.Map.find(Piece.id(p), M.terms) |> Term.rep_id;
```
> the decoration when the cursor is on a comma in a partial application are missing the f (it works correctly when on the parentheses):
> ![image](https://github.com/hazelgrove/hazel/assets/108375845/aded2285-dda7-4b0f-9c68-a608569ab5e6)

The issue I encountered while fixing this bug is that the `TermRanges` seems to be related to `Skel`. Even if I can break down and reconstruct `Term` in `MakeTerm`, I'm not sure how to do the same for the corresponding `Skel`. It appears that the creation of `Skel` doesn't take into account the information from `Term`.

Here is the solution that Haoxiang came up with. I'm not sure if it will have any impact on existing programs, but it appears to be the least disruptive solution to the codebase.

disconcision replied:
Interesting.... TermRanges are currently not working quite right in dev for both list literals and case rules, so it's not the end of the world if it doesn't work here. I don't really know the details of TermRanges so @dm0n3y will need to weigh in on whether this change is likely to be disruptive or not

dm0n3y replied:
looks good to me. in fact, probably should have been written this way originally (using the rep_id).

pigumar1 committed:
Merge branch 'adt-defs-after' into haz3l-partial-application-no-self-type

cyrus- reviewed (changes requested):
crashes when I try to turn on the langdocs color toggle at this state 
<img width="1601" alt="image" src="https://github.com/hazelgrove/hazel/assets/280638/fb48ca39-1632-433e-8f30-93df59a8a05c">

pigumar1 commented:
> crashes when I try to turn on the langdocs color toggle at this state <img alt="image" width="1601" src="https://user-images.githubusercontent.com/280638/257076125-fb48ca39-1632-433e-8f30-93df59a8a05c.png">

@cyrus- how should we highlight the arguments? We could highlight all of them as a whole, highlight each of them separately, or we don't highlight any of them.

cyrus- commented:
> > crashes when I try to turn on the langdocs color toggle at this state <img alt="image" width="1601" src="https://user-images.githubusercontent.com/280638/257076125-fb48ca39-1632-433e-8f30-93df59a8a05c.png">
> 
> @cyrus- how should we highlight the arguments? We could highlight all of them as a whole, highlight each of them separately, or we don't highlight any of them.

Ideally it would highlight the partially applied arguments in a different color from the other arguments, which can all have the same color. So the message can say "<regular arguments> are partially applied and <deferred arguments> are deferred".

cyrus- committed:
Merge branch 'adt-defs-after' into haz3l-partial-application-no-self-type

pigumar1 committed:
Code optimization + documentation (fancy!)

pigumar1 committed:
Merge branch 'haz3l-partial-application-no-self-type' of https://github.com/hazelgrove/hazel into haz3l-partial-application-no-self-type

pigumar1 commented on src/haz3lweb/view/LangDoc.re:
```diff
@@ -1902,6 +1915,50 @@ let get_doc =
+              ~deferred_id,
+            ),
+          )
+        | Colorings =>
```
![image](https://github.com/hazelgrove/hazel/assets/108375845/ff2e72e5-01d7-46f7-8eac-a253b7dd4b73)
Color highlighting is done though the implementation in `LangDoc.re` has become really hacky... It seems that each term should have a different color and this implementation circumvents `ColorSteps.get_color` to have all deferrals highlighted with the same color.

pigumar1 commented on src/haz3lweb/view/LangDoc.re:
```diff
@@ -1902,6 +1915,50 @@ let get_doc =
+      | DeferredAp(x, args) =>
+        let x_id = List.nth(x.ids, 0);
+        let supplied_id = (-1);
+        let deferred_id = (-2);
```
Hacky: the coloring ids are decoupled with terms, since there might be a case in which no arguments are supplied.

cyrus- committed:
Merge branch 'adt-defs-after' into haz3l-partial-application-no-self-type

cyrus- reviewed (changes requested):
- The lang docs message should "The deferred arguments can be applied in future applications." rather than "unsupplied arguments..."

- The example in the lang doc message should use a let binding rather than applying an anonymous function for clarity, and to avoid horizontal scrolling.

pigumar1 committed:
Update LangDocMessages.re

pigumar1 committed:
Merge branch 'adt-defs-after' into haz3l-partial-application-no-self-type

pigumar1 commented:
![image](https://github.com/hazelgrove/hazel/assets/108375845/8b84a915-3106-45c4-8d22-4f6920af2f1d)
How about this?

pigumar1 committed:
Merge branch 'adt-defs-after' into haz3l-partial-application-no-self-type

pigumar1 committed:
Small fixes

cyrus- commented:
looks good! 

@pigumar1 can you resolve the conflicts with the adts branch, @disconcision is going to take a look at the code after that

pigumar1 committed:
Merge branch 'adt-defs-after' into haz3l-partial-application-no-self-type

pigumar1 committed:
Merge branch 'dev' into haz3l-partial-application-no-self-type

disconcision commented on src/haz3lcore/statics/Statics.re:
```diff
@@ -160,6 +160,8 @@ and uexp_to_info_map =
   | EmptyHole => atomic(Just(Unknown(Internal)))
   | Triv => atomic(Just(Prod([])))
+  | Deferral(status_deferral) =>
+    add'(~self=IsDeferral(status_deferral), ~co_ctx=CoCtx.empty, m)
```
use atomic here instead of add'.
just call 'status_deferral' 'status', we know the deferral part from context.

disconcision commented on src/haz3lcore/statics/Statics.re:
```diff
@@ -236,6 +238,19 @@ and uexp_to_info_map =
+    let (fn, m) = go(~mode=fn_mode, fn, m);
+    let (ty_in, ty_out) = Typ.matched_arrow(fn.ty);
+    let (self: Self.exp, m, arg_co_ctx) = {
+      let length = List.length(args);
```
let's call this num_args

disconcision commented on src/haz3lcore/statics/Statics.re:
```diff
@@ -236,6 +238,19 @@ and uexp_to_info_map =
+    let fn_mode = Mode.of_ap(ctx, mode, UExp.ctr_name(fn));
+    let (fn, m) = go(~mode=fn_mode, fn, m);
+    let (ty_in, ty_out) = Typ.matched_arrow(fn.ty);
+    let (self: Self.exp, m, arg_co_ctx) = {
```
I don't think this needs to be a nested block

disconcision commented on src/haz3lcore/statics/Statics.re:
```diff
@@ -236,6 +238,19 @@ and uexp_to_info_map =
+      let length = List.length(args);
+      let ty_ins = Typ.matched_args(length, ty_in);
+      let self = Self.of_deferred_ap(args, ty_ins, ty_out);
+      let modes = Mode.of_deferred_ap_args(ty_ins, length);
```
Typ.matched_args and Mode.deferred_ap_args have their parameters in opposite order; let's go length-first for both, same as matched_prod

disconcision commented on src/haz3lcore/statics/TermBase.re:
```diff
@@ -97,12 +97,18 @@ and UExp: {
+  [@deriving (show({with_path: false}), sexp, yojson)]
+  type status_deferral =
+    | Used
+    | Unused;
```
Not obvious to me what these names connote... maybe something more like InAp and OutsideAp? and something like "deferral_position" instead of "status_deferral"?

disconcision commented on src/haz3lcore/statics/Mode.re:
```diff
@@ -131,3 +131,10 @@ let of_ap = (ctx, mode, ctr: option(Constructor.t)): t =>
+    List.length(ty_ins) == length
+      ? ty_ins : List.init(length, _ => (Unknown(Internal): Typ.t))
+  )
+  |> List.map(ty => Ana(ty));
```
no need for a separate map here, just add the Ana to the Unknown(Internal(

disconcision commented on src/haz3lcore/statics/Self.re:
```diff
@@ -35,7 +35,15 @@ type t =
+  | IsDeferral(Term.UExp.status_deferral)
+  | IsErroneousPartialAp(error_partial_ap)
+  | Common(t)
+and error_partial_ap =
```
and is mostly used for mutually recursive types... just define this type separately above exp

disconcision commented on src/haz3lcore/statics/Self.re:
```diff
@@ -35,7 +35,15 @@ type t =
+  | IsErroneousPartialAp(error_partial_ap)
+  | Common(t)
+and error_partial_ap =
+  | Meaningless
```
'Meaningless' needs a better name... maybe NoDefferedArgs?

disconcision commented on src/haz3lcore/statics/Self.re:
```diff
@@ -35,7 +35,15 @@ type t =
   | Free(Var.t)
-  | Common(t);
+  | IsDeferral(Term.UExp.status_deferral)
+  | IsErroneousPartialAp(error_partial_ap)
```
let's just call it BadPartialAp; for better or worse we already use Bad this way in the rest of the codebase

disconcision commented on src/haz3lcore/statics/Self.re:
```diff
@@ -85,6 +95,23 @@ let of_ctr = (ctx: Ctx.t, name: Constructor.t): t =>
+  let actual = List.length(args);
+  if (expected != actual) {
+    IsErroneousPartialAp(ArityMismatch({expected, actual}));
+  } else if (!List.exists(arg => !Term.UExp.is_deferral(arg), args)) {
```
eliminate the double negative... something like `List.for_all(is_deferral, args)`

disconcision commented on src/haz3lcore/statics/Self.re:
```diff
@@ -85,6 +95,23 @@ let of_ctr = (ctx: Ctx.t, name: Constructor.t): t =>
+      List.combine(args, ty_ins)
+      |> List.filter(((arg, _ty)) => Term.UExp.is_deferral(arg))
+      |> List.map(snd);
+    let ty_in = List.length(ty_ins) == 1 ? List.hd(ty_ins) : Prod(ty_ins);
```
This is basically the same as mk_tuple... might want to consider making this a Typ-level helper... not sure though your call.

disconcision commented on src/haz3lcore/statics/Info.re:
```diff
@@ -95,7 +97,9 @@ type ok_common =
 [@deriving (show({with_path: false}), sexp, yojson)]
-type ok_exp = ok_common;
+type ok_exp =
+  | AnaDeferralConsistent(Typ.t)
```
Can you remind me why we need this additional ok status? I don't think we need if for status_exp because we already know by construction in maketerm that every deferral inside a partial ap will be a Deferral(Used))

disconcision commented on src/haz3lcore/dynamics/Elaborator.re:
```diff
@@ -241,6 +242,47 @@ let rec dhexp_of_uexp =
+            mk_tuple(x => DHExp.Tuple(x), c_args),
+            mk_tuple(x => Typ.Prod(x), ty_args),
+          );
+          DHExp.Fun(pat, Arrow(ty_arg, ty_ret), Ap(c_fn, c_arg), None);
```
This is going to require some manual casting... worth talking about with Cyrus, it's kind of complicated. Here's an example that will trigger a casting exception when run (the "?" are holes):

`let f: ?  -> ?  = fun a,b -> a+ b in
let g = f(1,_) in
g(1)`

disconcision commented on src/haz3lweb/LangDocMessages.re:
```diff
@@ -1586,6 +1621,47 @@ let conapp_exp: form = {
+  };
+  let comma = comma_exp();
+  let or_exp = () =>
+    Example.mk_tile(Form.mk_infix("|", Rul, Precedence.or_), []);
```
this is rendering as red (sort inconsistency error) as this is not a valid hazel operator on its own... This needs to be a concrete example, not a schema... probably just drop the "e1 |" part and just go with the "_"

disconcision reviewed (changes requested):
Looking pretty solid. The only functional problem I could find is related to casting. See my comment in Elaborator for a problematic example. Casts are usually applied automatically but the way the rest of the code is set up right now, they need to be manually constructed in cases like this where the DHExp is manually constructed. I don't immediately see what the correct casting approach is; I'm going to defer to Cyrus on this one.

I'm also not totally clear on the functional role being played by AnaDeferralConsistent; see my comment there.

pigumar1 committed:
Code improvement

cyrus- commented:
@pigumar1 have you resolved all of @disconcision's comments above? if so, please mark them as resolved. marking as draft for now.

pigumar1 committed:
Code improvement 2

pigumar1 commented:
> @pigumar1 have you resolved all of @disconcision's comments above? if so, please mark them as resolved. marking as draft for now.

Because the dev branch has been updated, new bugs have been exposed. I am currently addressing them...

pigumar1 committed:
Got rid of the "or pattern" for the document

pigumar1 committed:
Update Elaborator.re

pigumar1 committed:
Merge remote-tracking branch 'origin/randIds' into haz3l-partial-application-no-self-type

pigumar1 committed:
"Desugaring" implementation

pigumar1 committed:
Reformatted

cyrus- commented:
@pigumar1 click Ready for review when this is ready again

pigumar1 committed:
Reverted desugaring implementation

pigumar1 committed:
Circumvented manual casting with desugaring

pigumar1 committed:
Merge remote-tracking branch 'origin/randIds' into haz3l-partial-application-no-self-type

pigumar1 committed:
weak head normalized

pigumar1 committed:
Make Typ.matched_args take ctx into account

pigumar1 committed:
Merge branch 'dev' into haz3l-partial-application-no-self-type

pigumar1 committed:
Fixed Typ.matched_arrow

pigumar1 committed:
Merge branch 'dev' into haz3l-partial-application-no-self-type

pigumar1 committed:
Merge branch 'dev' into haz3l-partial-application-no-self-type

pigumar1 committed:
Merge branch 'dev' into haz3l-partial-application-no-self-type

cyrus- commented:
@pigumar1 can you merge with the latest changes, should be small -- mostly be the langdocs rework. will prioritize getting it merged into `dev` after that.

pigumar1 committed:
Merge branch 'dev' into haz3l-partial-application-no-self-type

pigumar1 committed:
Documentation

pigumar1 committed:
Merge branch 'dev' into haz3l-partial-application-no-self-type

pigumar1 commented:
I added cases for `Deferral` and `DeferredAp `to `SyntaxTest.re`, but have no idea whether they are correct or not...

cyrus- committed:
merge dev and fix syntax tests

cyrus- committed:
remove commented out merge conflict

disconcision commented on src/haz3lcore/statics/Statics.re:
```diff
@@ -243,11 +243,11 @@ and uexp_to_info_map =
-      ~co_ctx=CoCtx.union([fn.co_ctx, arg.co_ctx]),
-      m,
-    );
+    let self: Self.t =
```
ideally break this out into Self.of_ap

disconcision reviewed (approved):
lgtm

disconcision commented:
actually I feel this error could be tightened up, or at least capitalization could be more consistent:
"Arity mismatched partial application: expected 2 arguments, got 1"
'partial application' is redundant; the label right before it says that.
how about "Arity mismatch: Expected 2 arguments, got 1"

cyrus- committed:
tweaks to messages
